stmhal/diskio.c

```diff
@@ -242,29 +242,64 @@ DRESULT disk_ioctl (
             break;
 #endif
 
-        case PD_USER:
-            if (MP_STATE_PORT(fs_user_mount) == NULL) {
+        case PD_USER: {
+            fs_user_mount_t *vfs = MP_STATE_PORT(fs_user_mount);
+            if (vfs == NULL) {
                 // nothing mounted
                 return RES_ERROR;
             }
-            switch (cmd) {
-                case CTRL_SYNC:
-                    if (MP_STATE_PORT(fs_user_mount)->sync[0] != MP_OBJ_NULL) {
-                        mp_call_method_n_kw(0, 0, MP_STATE_PORT(fs_user_mount)->sync);
+            if (vfs->u.old.count[1] == MP_OBJ_SENTINEL) {
+                // new protocol with ioctl
+                switch (cmd) {
+                    case CTRL_SYNC:
+                        vfs->u.ioctl[2] = MP_OBJ_NEW_SMALL_INT(0 /* SYNC */);
+                        mp_call_method_n_kw(1, 0, vfs->u.ioctl);
+                        return RES_OK;
+
+                    case GET_SECTOR_COUNT: {
+                        vfs->u.ioctl[2] = MP_OBJ_NEW_SMALL_INT(1 /* SEC_COUNT */);
+                        mp_obj_t ret = mp_call_method_n_kw(1, 0, vfs->u.ioctl);
+                        *((DWORD*)buff) = mp_obj_get_int(ret);
+                        return RES_OK;
                     }
-                    return RES_OK;
 
-                case GET_BLOCK_SIZE:
-                    *((DWORD*)buff) = 1; // high-level sector erase size in units of the small (512) bl
-                    return RES_OK;
+                    case GET_SECTOR_SIZE: {
+                        vfs->u.ioctl[2] = MP_OBJ_NEW_SMALL_INT(2 /* SEC_SIZE */);
+                        mp_obj_t ret = mp_call_method_n_kw(1, 0, vfs->u.ioctl);
+                        *((WORD*)buff) = mp_obj_get_int(ret);
```

Comment:
@pfalcon in your PR you had this as a DWORD, but I think it should be a WORD pointer.